test/reduce.jl

```diff
@@ -613,6 +613,7 @@ test18695(r) = sum( t^2 for t in r )
 @test_throws str -> ( occursin("reducing over an empty", str) &&
                       occursin("consider supplying `init`", str) &&
                      !occursin("or defining", str)) test18695(Any[])
+@test_throws MethodError test18695(Any[])
```

Comment:
I blindly went to add some tests here as I started reviewing this, and I wrote the following...

```suggestion
@test_throws MethodError test18695(Any[])

@testset "issue #52457: pairwise reduction of iterators" begin
    @test foldl(+, Iterators.repeated(Float16(1.0), 10000)) == maxintfloat(Float16)
    @test mapfoldl(x->x^2, +, Iterators.repeated(Float16(1.0), 10000)) == maxintfloat(Float16)
    @test foldr(+, Iterators.repeated(Float16(1.0), 10000)) == maxintfloat(Float16)
    @test mapfoldr(x->x^2, +, Iterators.repeated(Float16(1.0), 10000)) == maxintfloat(Float16)

    @test reduce(+, Iterators.repeated(Float16(1.0), 10000)) ≈ 10000
    @test mapreduce(x->x^2, +, Iterators.repeated(Float16(1.0), 10000)) ≈ 10000

    @test_throws ArgumentError reduce((x,y)->x+y, Iterators.repeated(Float16(1.0), 0))
    @test_throws ArgumentError mapreduce(identity, (x,y)->x+y, Iterators.repeated(Float16(1.0), 0))

    @test reduce(+, Iterators.repeated(Float16(1.0), 0); init=Float16(0)) === Float16(0)
    @test mapreduce(identity, +, Iterators.repeated(Float16(1.0), 0); init=Float16(0)) === Float16(0)
    @test reduce(+, Iterators.repeated(Float16(1.0), 10000); init=Float16(0)) ≈ 10000
    @test mapreduce(identity, +, Iterators.repeated(Float16(1.0), 10000); init=Float16(0)) ≈ 10000
    @test reduce((x,y)->x+y, Iterators.repeated(Float16(1.0), 0); init=Float16(0)) === Float16(0)
    @test mapreduce(x->x^2, (x,y)->x+y, Iterators.repeated(Float16(1.0), 0); init=Float16(0)) === Float16(0)
    @test reduce((x,y)->x+y, Iterators.repeated(Float16(1.0), 10000); init=Float16(0)) ≈ 10000
    @test mapreduce(x->x^2, (x,y)->x+y, Iterators.repeated(Float16(1.0), 10000); init=Float16(0)) ≈ 10000
end
```

But that fails, because supplying init pushes you back into foldl land.  I see you've considered this... but it still feels pretty unexpected.